[lit-html] Add DEV_MODE error if duplicate attribute bindings are encountered

AndrewJakubowicz opened the pull request:
Fixes: https://github.com/lit/lit/issues/4194

### Context

A lot of the context is in the issue, but in short, if an attribute binding is duplicated, such as in the following example:

```html
<textarea 
      ?disabled=${this.disabled} 
      ?disabled=${this.disabled} 
      placeholder=${this.placeholder}
></textarea>
```

The result is that Lit parts get out of sync with the values (because attributes are de-duped by the browser).

### Fix

I've implemented this as a DEV_MODE error, but I'm open to making it a DEV_MODE warning.

The following template: `<input ?disabled=${true} ?disabled=${false}>`, will throw:

```
Detected duplicate attribute bindings of attribute: '?disabled', in the template: `<input ?disabled=${...} ?disabled=${...}>`
```

justinfagnani reviewed (commented):
lgtm!!

justinfagnani commented:
I think an error is correct here. The template will throw at rendering time anyway.

AndrewJakubowicz commented:
Note: Converted to draft because this collapses duplicate attributes across individual tags which is invalid behavior.

Added a test case: `<input ?disabled=${...}><input ?disabled=${...}>`. This should not error (and it does with the current naive implementation).

AndrewJakubowicz commented:
I fixed the implementation to be correct. Throw if there are attribute names that were not considered for template preparation (detects attribute mismatch).

Adding the attribute that caused the attribute mismatch is going to be tricky as duplicate attribute names can exist across multiple tags. I think this implementation is strictly an improvement, and we provide the template in the error that caused the issue. This will reduce debugging time for user's who encounter this error.

augustjk reviewed (approved):
Ooohhh very clever to check that we've consumed everything from `attrNames`!

sorvell commented on packages/lit-html/src/lit-html.ts:
```diff
@@ -1036,6 +1036,26 @@ class Template {
+      // If there was a duplicate attribute on a tag, then when the tag is
+      // parsed into an element the attribute gets de-duplicated. We can detect
+      // this mismatch if we haven't precisely consumed every attribute name
+      // when preparing the template.
```
```suggestion
      // when preparing the template. This works because `attrNames` is built from the template string and `attrNameIndex` comes from processing the resulting DOM.
```
